19.03-rc/dind-rootless/Dockerfile

```diff
@@ -0,0 +1,81 @@
+FROM docker:19.03-rc-dind
+
+# busybox "ip" is insufficient:
+#   [rootlesskit:child ] error: executing [[ip tuntap add name tap0 mode tap] [ip link set tap0 address 02:50:00:00:00:01]]: exit status 1
+RUN apk add --no-cache iproute2
+
+# "/run/user/UID" will be used by default as the value of XDG_RUNTIME_DIR
+RUN mkdir /run/user && chmod 1777 /run/user
+
+# create a default user preconfigured for running rootless dockerd
+RUN set -eux; \
+	adduser -h /home/rootless -g 'Rootless' -D -u 1000 rootless; \
+	echo 'rootless:100000:65536' >> /etc/subuid; \
+	echo 'rootless:100000:65536' >> /etc/subgid
+
+RUN set -eux; \
+	\
+# this "case" statement is generated via "update.sh"
+	apkArch="$(apk --print-arch)"; \
+	case "$apkArch" in \
+# amd64
+		x86_64) dockerArch='x86_64' ;; \
+# arm32v6
+		armhf) dockerArch='armel' ;; \
+# arm32v7
+		armv7) dockerArch='armhf' ;; \
+# arm64v8
+		aarch64) dockerArch='aarch64' ;; \
+		*) echo >&2 "error: unsupported architecture ($apkArch)"; exit 1 ;;\
+	esac; \
+	\
+	if ! wget -O rootless.tgz "https://download.docker.com/linux/static/${DOCKER_CHANNEL}/${dockerArch}/docker-rootless-extras-${DOCKER_VERSION}.tgz"; then \
+		echo >&2 "error: failed to download 'docker-rootless-extras-${DOCKER_VERSION}' from '${DOCKER_CHANNEL}' for '${dockerArch}'"; \
+		exit 1; \
+	fi; \
+	\
+	tar --extract \
+		--file rootless.tgz \
+		--strip-components 1 \
+		--directory /usr/local/bin/ \
+		'docker-rootless-extras/vpnkit' \
+	; \
+	rm rootless.tgz; \
+	\
+# we download/build rootlesskit separately to get a newer release
+#	rootlesskit --version; \
+	vpnkit --version
+
+# https://github.com/rootless-containers/rootlesskit/releases
+ENV ROOTLESSKIT_VERSION 0.6.0
+
+RUN set -eux; \
+	apk add --no-cache --virtual .rootlesskit-build-deps \
+		go \
```

Comment:
can we use multi-stage build?

Reply:
Technically yes, but I'm not sure I understand what benefit that would bring?  Adding and removing `go` here is pretty trivial (and makes sure we comply with https://github.com/docker-library/faq#multi-stage-builds).

Reply:
BuildKit can build images faster when there is concurrency across stages.

Reply:
BTW you don't need to build RootlessKit v0.6.0 because it is included in the nightly binary: https://download.docker.com/linux/static/nightly/x86_64/

Reply:
Build time is pretty negligible here (~6.6s on my machine):
```dockerfile
FROM alpine:3.10

SHELL ["time", "sh", "-c"]

# https://github.com/rootless-containers/rootlesskit/releases
ENV ROOTLESSKIT_VERSION 0.6.0

RUN set -eux; \
	apk add --no-cache --virtual .rootlesskit-build-deps \
		go \
		libc-dev \
	; \
	wget -O rootlesskit.tgz "https://github.com/rootless-containers/rootlesskit/archive/v${ROOTLESSKIT_VERSION}.tar.gz"; \
	export GOPATH='/go'; mkdir "$GOPATH"; \
	mkdir -p "$GOPATH/src/github.com/rootless-containers/rootlesskit"; \
	tar --extract --file rootlesskit.tgz --directory "$GOPATH/src/github.com/rootless-containers/rootlesskit" --strip-components 1; \
	rm rootlesskit.tgz; \
	go build -o /usr/local/bin/rootlesskit github.com/rootless-containers/rootlesskit/cmd/rootlesskit; \
	go build -o /usr/local/bin/rootlesskit-docker-proxy github.com/rootless-containers/rootlesskit/cmd/rootlesskit-docker-proxy; \
	rm -rf "$GOPATH"; \
	apk del --no-network .rootlesskit-build-deps; \
	rootlesskit --version
```

```console
$ docker build .
Sending build context to Docker daemon  15.87kB
Step 1/4 : FROM alpine:3.10
 ---> b7b28af77ffe
Step 2/4 : SHELL ["time", "sh", "-c"]
 ---> Using cache
 ---> 9723a1c87c6d
Step 3/4 : ENV ROOTLESSKIT_VERSION 0.6.0
 ---> Using cache
 ---> e8f79cbc7791
Step 4/4 : RUN set -eux; 	apk add --no-cache --virtual .rootlesskit-build-deps 		go 		libc-dev 	; 	wget -O rootlesskit.tgz "https://github.com/rootless-containers/rootlesskit/archive/v${ROOTLESSKIT_VERSION}.tar.gz"; 	export GOPATH='/go'; mkdir "$GOPATH"; 	mkdir -p "$GOPATH/src/github.com/rootless-containers/rootlesskit"; 	tar --extract --file rootlesskit.tgz --directory "$GOPATH/src/github.com/rootless-containers/rootlesskit" --strip-components 1; 	rm rootlesskit.tgz; 	go build -o /usr/local/bin/rootlesskit github.com/rootless-containers/rootlesskit/cmd/rootlesskit; 	go build -o /usr/local/bin/rootlesskit-docker-proxy github.com/rootless-containers/rootlesskit/cmd/rootlesskit-docker-proxy; 	rm -rf "$GOPATH"; 	apk del --no-network .rootlesskit-build-deps; 	rootlesskit --version
 ---> Running in 2166516f6dc2
+ apk add --no-cache --virtual .rootlesskit-build-deps go libc-dev
fetch http://dl-cdn.alpinelinux.org/alpine/v3.10/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.10/community/x86_64/APKINDEX.tar.gz
...
(14/14) Installing .rootlesskit-build-deps (20190801.213242)
Executing busybox-1.30.1-r2.trigger
OK: 391 MiB in 28 packages
+ wget -O rootlesskit.tgz https://github.com/rootless-containers/rootlesskit/archive/v0.6.0.tar.gz
...
rootlesskit.tgz      100% |********************************| 1519k  0:00:00 ETA
+ export 'GOPATH=/go'
+ mkdir /go
+ mkdir -p /go/src/github.com/rootless-containers/rootlesskit
+ tar --extract --file rootlesskit.tgz --directory /go/src/github.com/rootless-containers/rootlesskit --strip-components 1
+ rm rootlesskit.tgz
+ go build -o /usr/local/bin/rootlesskit github.com/rootless-containers/rootlesskit/cmd/rootlesskit
+ go build -o /usr/local/bin/rootlesskit-docker-proxy github.com/rootless-containers/rootlesskit/cmd/rootlesskit-docker-proxy
+ rm -rf /go
+ apk del --no-network .rootlesskit-build-deps
...
OK: 6 MiB in 14 packages
+ rootlesskit --version
rootlesskit version 0.6.0
real	0m 6.65s
user	0m 6.46s
sys	0m 1.41s
Removing intermediate container 2166516f6dc2
 ---> f23d61b27573
Successfully built f23d61b27573
```

Reply:
Nightly builds have no way to correlate that the version `0.0.0-[date]` contains version `0.6.0` of `rootlesskit`; using the upstream `rootlesskit` releases gives better control of when it is updated and users more visibility into its provenance.

Side note: if `rootlesskit` had binary releases with signatures we would definitely be open to using them.